Sync/BatchingClient.swift

```diff
@@ -0,0 +1,218 @@
+import Alamofire
+import Shared
+import XCGLogger
+import Deferred
+
+public class SerializeRecordFailure<T: CleartextPayloadJSON>: MaybeErrorType {
+    public let record: Record<T>
+
+    public var description: String {
+        return "Failed to serialize record: \(record)"
+    }
+
+    public init(record: Record<T>) {
+        self.record = record
+    }
+}
+
+private let log = Logger.syncLogger
+
+private typealias UploadRecord = (guid: GUID, payload: String, size: Int)
+private typealias DeferredResponse = Deferred<Maybe<StorageResponse<POSTResult>>>
+
+typealias BatchUploadFunction = (lines: [String], ifUnmodifiedSince: Timestamp?, queryParams: [NSURLQueryItem]?) -> Deferred<Maybe<StorageResponse<POSTResult>>>
+
+public class Sync15BatchClient<T: CleartextPayloadJSON> {
+    private(set) var ifUnmodifiedSince: Timestamp?
+
+    private let config: InfoConfiguration
+    private let uploader: BatchUploadFunction
+    private let serializeRecord: (Record<T>) -> String?
+
+    private let commitParam = NSURLQueryItem(name: "commit", value: "true")
+
+    private var batchToken: BatchToken?
+
+    // Keep track of the limits of a single batch
+    private var totalBytes: ByteCount = 0
+    private var totalRecords: Int = 0
+
+    // Keep track of the limits of a single POST
+    private var postBytes: ByteCount = 0
+    private var postRecords: Int = 0
+
+    private var records = [UploadRecord]()
+
+    private var onCollectionUploaded: (POSTResult, Timestamp?) -> DeferredTimestamp
+
+    private func batchQueryParamWithValue(value: String) -> NSURLQueryItem {
+        return NSURLQueryItem(name: "batch", value: value)
+    }
+
+    init(config: InfoConfiguration, ifUnmodifiedSince: Timestamp? = nil, serializeRecord: (Record<T>) -> String?,
+         uploader: BatchUploadFunction, onCollectionUploaded: (POSTResult, Timestamp?) -> DeferredTimestamp) {
+        self.config = config
+        self.ifUnmodifiedSince = ifUnmodifiedSince
+        self.uploader = uploader
+        self.serializeRecord = serializeRecord
+
+        self.onCollectionUploaded = onCollectionUploaded
+    }
+
+    public func endBatch() -> Success {
+        guard !records.isEmpty else {
+            return succeed()
+        }
+
+        if let token = self.batchToken {
+            return commitBatch(token) >>> succeed
+        } else {
+            let lines = self.freezePost()
+            let queryParams = [
+                batchQueryParamWithValue("true"),
+                NSURLQueryItem(name: "commit", value: "true")
+            ]
+            return self.uploader(lines: lines, ifUnmodifiedSince: self.ifUnmodifiedSince, queryParams: queryParams)
+                >>== effect(moveForward)
+                >>> succeed
+        }
+    }
+
+    public func addRecords(records: [Record<T>]) -> Success {
+        guard !records.isEmpty else {
+            return succeed()
+        }
+
+        do {
+            let serialized = try records.map { try serialize($0) }
```

Comment:
Worth adding a comment to the effect of some of my review comments: we eagerly serialize all records before we queue them, because if we're going to fail we want to do it early.